chain/epoch-manager/src/lib.rs

```diff
@@ -800,11 +810,57 @@ impl EpochManager {
                 if self.is_next_block_in_next_epoch(&block_info)? {
                     self.finalize_epoch(&mut store_update, &block_info, &current_hash, rng_seed)?;
                 }
+
+                if self.is_next_block_in_next_epoch(&prev_block_info)? {
+                    self.update_epoch_sync_proof(&block_info)?;
+                }
             }
         }
         Ok(store_update)
     }
 
+    /// We call this function on the first block of epoch T. We update the epoch sync proof to that
+    /// of epoch T-2. We pass the last_block_hash of epoch T-2 to extend the epoch sync proof.
+    ///
+    /// Any new node doing epoch sync needs to have at least `transaction_validity_period` number of
+    /// block headers to validate transactions.
+    /// Currently, `transaction_validity_period` is set to ~2 epochs worth of blocks, which is why
+    /// we update the epoch sync proof to that of epoch T-2 here.
+    ///
+    /// In the future, if `transaction_validity_period` were to increase, we would need to update
+    /// this function.
+    fn update_epoch_sync_proof(&self, first_block_info: &BlockInfo) -> Result<(), EpochError> {
+        // pe -> previous_epoch
+        // ppe -> previous_previous_epoch
+        let last_block_hash_in_pe = first_block_info.prev_hash();
+        let last_block_info_in_pe = self.store.get_block_info(last_block_hash_in_pe)?;
+        let first_block_hash_in_ppe = last_block_info_in_pe.epoch_first_block();
+        let first_block_info_in_ppe = self.store.get_block_info(first_block_hash_in_ppe)?;
+        let last_block_hash_in_ppe = first_block_info_in_ppe.prev_hash();
+
+        // Assert for transaction_validity_period.
+        // First block of the target epoch should be at lease `transaction_validity_period` blocks
+        // behind the current head
+        let last_block_info_in_ppe = self.store.get_block_info(last_block_hash_in_ppe)?;
+        let first_block_hash_in_ppe = last_block_info_in_ppe.epoch_first_block();
+        let first_block_info_in_ppe = self.store.get_block_info(first_block_hash_in_ppe)?;
+        if !first_block_info_in_ppe.is_genesis() {
+            let being_height = first_block_info_in_ppe.height();
+            let end_height = first_block_info.height();
+            assert!(
+                being_height + self.transaction_validity_period <= end_height,
```

Comment:
So turns out a LOT of tests are failing as `transaction_validity_period <= epoch_length * 2` doesn't really hold in a lot of tests.

I'm working on a separate PR to fix this, and while at it, I got the better idea of testing this directly in the constructor of epoch manager instead of checking every time here.

Both `epoch_length` and `transaction_validity_period` are anyway only in the epoch config and genesis and are effective constants.